src/structs/lepton_file_reader.rs

```diff
@@ -0,0 +1,679 @@
+        // data to store any extra RST codes or whatever else might be at the end of the file.
+        if lh.rst_err.len() > 0 {
+            let mut markers = Vec::new();
+
+            let cumulative_reset_markers = if lh.jpeg_header.rsti != 0 {
+                ((lh.jpeg_header.mcuh * lh.jpeg_header.mcuv) - 1) / lh.jpeg_header.rsti
+            } else {
+                0
+            } as u8;
+
+            for i in 0..lh.rst_err[0] as u8 {
+                let rst = (jpeg_code::RST0 + ((cumulative_reset_markers + i) & 7)) as u8;
+                markers.push(0xFF);
+                markers.push(rst);
+            }
+
+            let expected_total_length = results.iter().map(|x| x.len()).sum::<usize>()
+                + lh.garbage_data.len()
+                + (lh.raw_jpeg_header.len() - lh.raw_jpeg_header_read_index);
+
+            if expected_total_length < lh.jpeg_file_size as usize {
+                // figure out how much extra space we have, since C++ files can have
+                // more restart markers than there is space to fit them
+                let space_for_markers = min(
+                    markers.len(),
+                    lh.jpeg_file_size as usize - expected_total_length,
+                );
+
+                markers.resize(space_for_markers, 0);
+
+                results.push(markers);
+            }
+        }
+
+        Ok(DecoderState::AppendTrailer(results))
+    }
+
+    fn process_progressive(
+        lh: &mut LeptonHeader,
+        enabled_features: &EnabledFeatures,
+        mut image_segments: Vec<Vec<BlockBasedImage>>,
+    ) -> Result<DecoderState> {
+        let num_components = image_segments[0].len();
+        let mut merged = Vec::new();
+        for i in 0..num_components {
+            merged.push(BlockBasedImage::merge(&mut image_segments, i));
+        }
+
+        let mut header = Vec::new();
+        header.write_all(&SOI)?;
+        header
+            .write_all(&lh.raw_jpeg_header[0..lh.raw_jpeg_header_read_index])
+            .context(here!())?;
+
+        let mut results = Vec::new();
+        results.push(header);
+
+        loop {
+            // progressive JPEG consists of scans followed by headers
+            let scan =
+                jpeg_write_entire_scan(&merged[..], &JPegEncodingInfo::new(lh)).context(here!())?;
+            results.push(scan);
+
+            // read the next headers (DHT, etc) while mirroring it back to the writer
+            let old_pos = lh.raw_jpeg_header_read_index;
+            let result = lh
+                .advance_next_header_segment(enabled_features)
+                .context(here!())?;
+
+            results.push(lh.raw_jpeg_header[old_pos..lh.raw_jpeg_header_read_index].to_vec());
+
+            if !result {
+                break;
+            }
+
+            // advance to next scan
+            lh.scnc += 1;
+        }
+
+        Ok(DecoderState::AppendTrailer(results))
+    }
+
+    fn process_cmp(
```

Comment:
@mcroomp I am bit lost. I understand that this method should run decompression only on a specific chunk of data, but I do not see how it is achieved. Could you please explain?

Reply:
The caller uses inbuffer.take to make sure that there is a buffer of this size available.